File: 2020/01.rb
#!/usr/bin/env ruby
numbers = File.readlines("#{__dir__}/01.txt").map(&:to_i)

target = 2020
indices = Hash[(0..numbers.size).zip numbers].invert
numbers.each_with_index do |number, i|
  diff = target - number

  if indices.has_key?(diff) && indices[diff] != i # Also check we're not using the same expenditure twice
    puts "Part 1: #{number * diff}"
    break
  end
end

indices = {} # Maps from sum of two numbers -> indices of the operands used
numbers.each_with_index do |number, i|
  for j in 0..i - 1
    indices[number + numbers[j]] = [i, j]
  end
end
numbers.each do |number|
  diff = target - number
  if indices.has_key?(diff) && !indices[diff].include?(number) # Also check we're not using the same expenditure twice
    puts "Part 2: #{number * indices[diff].map{ |x| numbers[x] }.inject(:*)}"
    break
  end
end

File: 2015/01.php
#!/usr/bin/env php
<?php
$floor = 0;
$part2 = false;
foreach (str_split(rtrim(file_get_contents(__DIR__ . '/01.txt'))) as $i => $c) {
    if ($c == '(') {
        $floor++;
    } else {
        $floor--;
        if (!$part2 && $floor < 0) {
            $part2 = $i + 1;
        }
    }
}

echo "Part 1: {$floor}" . PHP_EOL;
echo "Part 2: {$part2}" . PHP_EOL;
